File: docs/virual man/demo/java/CreateAssetGroup_Demo (1).java
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.*;

/**
 * Copyright (year) Beijing Volcano Engine Technology Ltd.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class CreateAssetGroup_Demo {

    private static final BitSet URLENCODER = new BitSet(256);

    private static final String CONST_ENCODE = "0123456789ABCDEF";
    public static final Charset UTF_8 = StandardCharsets.UTF_8;

    private final String region;
    private final String service;
    private final String schema;
    private final String host;
    private final String path;
    private final String ak;
    private final String sk;

    static {
        int i;
        for (i = 97; i <= 122; ++i) {
            URLENCODER.set(i);
        }

        for (i = 65; i <= 90; ++i) {
            URLENCODER.set(i);
        }

        for (i = 48; i <= 57; ++i) {
            URLENCODER.set(i);
        }
        URLENCODER.set('-');
        URLENCODER.set('_');
        URLENCODER.set('.');
        URLENCODER.set('~');
    }

    /**
     * 构造函数，初始化必要的请求参数
     */
    public CreateAssetGroup_Demo(String region, String service, String schema, String host, String path, String ak, String sk) {
        this.region = region;
        this.service = service;
        this.host = host;
        this.schema = schema;
        this.path = path;
        this.ak = ak;
        this.sk = sk;
    }

    /**
     * 主程序执行入口，用于演示创建 Asset Group 的 API 调用
     */
    public static void main(String[] args) throws Exception {
        String AccessKeyID = "";
        String SecretAccessKey = "";
        // 请求地址
        String endpoint = "ark.cn-beijing.volcengineapi.com";
        String path = "/"; // 路径，不包含 Query
        // 请求接口信息
        String service = "ark";
        String region = "cn-beijing";
        String schema = "https";
        CreateAssetGroup_Demo sign = new CreateAssetGroup_Demo(region, service, schema, endpoint, path, AccessKeyID, SecretAccessKey);

        String action = "CreateAssetGroup";
        String version = "2024-01-01";

        Date date = new Date();
        HashMap<String, String> queryMap = new HashMap<String,String>() {{
        }};

        String body = "{\"Name\": \"test-group\", \"Description\": \"test\", \"GroupType\": \"AIGC\"}";

        System.out.println("==================================================");
        System.out.println("正在发起请求: Action=" + action + ", Version=" + version);
        System.out.println("请求体: " + body);
        System.out.println("==================================================");

        sign.doRequest("POST", queryMap, body.getBytes(), date, action, version);
    }

    /**
     * 封装对 OpenAPI 的签名计算和请求调用过程
     * 
     * @param method HTTP 请求方法
     * @param queryList URL 查询参数字典
     * @param body HTTP 请求体字节流
     * @param date 请求发起的时间
     * @param action API Action 名称
     * @param version API 版本
     */
    public void doRequest(String method, Map<String, String> queryList, byte[] body,
                          Date date, String action, String version) throws Exception {
        if (body == null) {
            body = new byte[0];
        }
        String xContentSha256 = hashSHA256(body);
        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd'T'HHmmss'Z'");
        sdf.setTimeZone(TimeZone.getTimeZone("GMT"));
        String xDate = sdf.format(date);
        String shortXDate = xDate.substring(0, 8);
        String contentType = "application/json";

        // Header 必须按照字母序排列
        String signHeader = "content-type;host;x-content-sha256;x-date";

        SortedMap<String, String> realQueryList = new TreeMap<>(queryList);
        realQueryList.put("Action", action);
        realQueryList.put("Version", version);
        StringBuilder querySB = new StringBuilder();
        for (String key : realQueryList.keySet()) {
            querySB.append(signStringEncoder(key)).append("=").append(signStringEncoder(realQueryList.get(key))).append("&");
        }
        querySB.deleteCharAt(querySB.length() - 1);

        // CanonicalRequest 中的 headers 也必须按照字母序排列
        String canonicalStringBuilder = method + "\n" + path + "\n" + querySB + "\n" +
                "content-type:" + contentType + "\n" +
                "host:" + host + "\n" +
                "x-content-sha256:" + xContentSha256 + "\n" +
                "x-date:" + xDate + "\n" +
                "\n" +
                signHeader + "\n" +
                xContentSha256;

        String hashcanonicalString = hashSHA256(canonicalStringBuilder.getBytes());
        String credentialScope = shortXDate + "/" + region + "/" + service + "/request";
        String signString = "HMAC-SHA256" + "\n" + xDate + "\n" + credentialScope + "\n" + hashcanonicalString;

        byte[] signKey = genSigningSecretKeyV4(sk, shortXDate, region, service);
        String signature = HexFormat.of().formatHex(hmacSHA256(signKey, signString));

        URL url = new URL(schema + "://" + host + path + "?" + querySB);

        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod(method);
        conn.setRequestProperty("Host", host);
        conn.setRequestProperty("X-Date", xDate);
        conn.setRequestProperty("X-Content-Sha256", xContentSha256);
        conn.setRequestProperty("Content-Type", contentType);
        conn.setRequestProperty("Authorization", "HMAC-SHA256" +
                " Credential=" + ak + "/" + credentialScope +
                ", SignedHeaders=" + signHeader +
                ", Signature=" + signature);
        if (!Objects.equals(conn.getRequestMethod(), "GET")) {
            conn.setDoOutput(true);
            OutputStream os = conn.getOutputStream();
            os.write(body);
            os.flush();
            os.close();
        }
        conn.connect();

        int responseCode = conn.getResponseCode();

        InputStream is;
        if (responseCode >= 200 && responseCode < 300) {
            is = conn.getInputStream();
        } else {
            is = conn.getErrorStream();
        }
        String responseBody = new String(is.readAllBytes());
        is.close();

        System.out.println("响应状态码: " + responseCode);
        System.out.println("响应内容:\n" + responseBody);
    }

    /**
     * 规范化查询参数（URL 编码）
     * 
     * @param source 原始字符串
     * @return 编码后的字符串
     */
    private String signStringEncoder(String source) {
        if (source == null) {
            return null;
        }
        StringBuilder buf = new StringBuilder(source.length());
        ByteBuffer bb = UTF_8.encode(source);
        while (bb.hasRemaining()) {
            int b = bb.get() & 255;
            if (URLENCODER.get(b)) {
                buf.append((char) b);
            } else if (b == 32) {
                buf.append("%20");
            } else {
                buf.append("%");
                char hex1 = CONST_ENCODE.charAt(b >> 4);
                char hex2 = CONST_ENCODE.charAt(b & 15);
                buf.append(hex1);
                buf.append(hex2);
            }
        }

        return buf.toString();
    }

    /**
     * 使用 SHA256 算法计算内容的哈希值
     * 
     * @param content 待哈希的字节流
     * @return 哈希后的十六进制字符串
     */
    public static String hashSHA256(byte[] content) throws Exception {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");

            return HexFormat.of().formatHex(md.digest(content));
        } catch (Exception e) {
            throw new Exception(
                    "Unable to compute hash while signing request: "
                            + e.getMessage(), e);
        }
    }

    /**
     * 使用 HMAC-SHA256 算法生成签名
     * 
     * @param key 签名的密钥
     * @param content 待签名的内容
     * @return 签名后的字节流
     */
    public static byte[] hmacSHA256(byte[] key, String content) throws Exception {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key, "HmacSHA256"));
            return mac.doFinal(content.getBytes());
        } catch (Exception e) {
            throw new Exception(
                    "Unable to calculate a request signature: "
                            + e.getMessage(), e);
        }
    }

    /**
     * 生成 V4 版本的签名密钥
     * 
     * @param secretKey 用户的 SecretAccessKey
     * @param date 日期字符串
     * @param region 区域
     * @param service 服务名称
     * @return 生成的签名密钥字节流
     */
    private byte[] genSigningSecretKeyV4(String secretKey, String date, String region, String service) throws Exception {
        byte[] kDate = hmacSHA256((secretKey).getBytes(), date);
        byte[] kRegion = hmacSHA256(kDate, region);
        byte[] kService = hmacSHA256(kRegion, service);
        return hmacSHA256(kService, "request");
    }
}

File: docs/virual man/demo/python/CreateAssetGroup_Demo (1).py
# coding:utf-8
import datetime
import hashlib
import hmac
import json
from urllib.parse import quote
import requests

# ===========================================================================
# 基础配置信息
# ===========================================================================
# 以下参数视服务不同而不同，一个服务内通常是一致的，不同接口注意修改action
SERVICE = "ark"
ACTION = "CreateAssetGroup"
VERSION = "2024-01-01"
REGION = "cn-beijing"
HOST = "ark.cn-beijing.volcengineapi.com"
CONTENT_TYPE = "application/json"
PATH = "/"

# 请求的凭证，从IAM或者STS服务中获取
AK = ""
SK = ""

def norm_query(params: dict) -> str:
    """
    规范化查询参数（Query String）。
    将字典类型的参数按照 key 的字典序排序，并进行 URL 编码。
    
    :param params: 查询参数字典
    :return: 规范化后的查询字符串
    """
    query = ""
    for key in sorted(params.keys()):
        if isinstance(params[key], list):
            for k in params[key]:
                query += quote(key, safe="-_.~") + "=" + quote(str(k), safe="-_.~") + "&"
        else:
            query += quote(key, safe="-_.~") + "=" + quote(str(params[key]), safe="-_.~") + "&"
    return query[:-1].replace("+", "%20")

def hmac_sha256(key: bytes, content: str) -> bytes:
    """
    使用 HMAC-SHA256 算法生成签名。
    
    :param key: 签名的密钥
    :param content: 待签名的内容
    :return: 签名后的字节流
    """
    return hmac.new(key, content.encode("utf-8"), hashlib.sha256).digest()

def hash_sha256(content: str) -> str:
    """
    使用 SHA256 算法计算内容的哈希值。
    
    :param content: 待哈希的内容
    :return: 哈希后的十六进制字符串
    """
    return hashlib.sha256(content.encode("utf-8")).hexdigest()

def request_api(method: str, date: datetime.datetime, query: dict, header: dict, ak: str, sk: str, body: str = "") -> requests.Response:
    """
    封装对 OpenAPI 的签名计算和请求调用过程。
    
    :param method: HTTP 请求方法，如 "GET", "POST"
    :param date: 请求发起的时间 (UTC 时间)
    :param query: URL 查询参数字典
    :param header: HTTP 请求头字典
    :param ak: Access Key ID
    :param sk: Secret Access Key
    :param body: HTTP 请求体 (默认为空字符串)
    :return: requests.Response 响应对象
    """
    # 1. 初始化身份证明结构体
    credential = {
        "access_key_id": ak,
        "secret_access_key": sk,
        "service": SERVICE,
        "region": REGION,
    }
    
    # 2. 初始化请求结构体
    # 合并基础 query 参数 (Action, Version) 和用户传入的 query 参数
    final_query = {"Action": ACTION, "Version": VERSION}
    if query:
        final_query.update(query)
        
    request_param = {
        "body": body if body else "",
        "host": HOST,
        "path": PATH,
        "method": method.upper(),
        "content_type": CONTENT_TYPE,
        "date": date,
        "query": final_query,
    }

    # 3. 准备签名的基础信息
    x_date = request_param["date"].strftime("%Y%m%dT%H%M%SZ")
    short_x_date = x_date[:8]
    x_content_sha256 = hash_sha256(request_param["body"])
    
    sign_result = {
        "Host": request_param["host"],
        "X-Content-Sha256": x_content_sha256,
        "X-Date": x_date,
        "Content-Type": request_param["content_type"],
    }
    
    signed_headers_str = ";".join(["content-type", "host", "x-content-sha256", "x-date"])
    
    # 4. 构造规范化请求串 (CanonicalRequest)
    canonical_request_str = "\n".join([
        request_param["method"],
        request_param["path"],
        norm_query(request_param["query"]),
        "\n".join([
            "content-type:" + request_param["content_type"],
            "host:" + request_param["host"],
            "x-content-sha256:" + x_content_sha256,
            "x-date:" + x_date,
        ]),
        "",
        signed_headers_str,
        x_content_sha256,
    ])
    
    hashed_canonical_request = hash_sha256(canonical_request_str)
    credential_scope = "/".join([short_x_date, credential["region"], credential["service"], "request"])
    
    # 5. 构造待签名字符串 (StringToSign)
    string_to_sign = "\n".join(["HMAC-SHA256", x_date, credential_scope, hashed_canonical_request])
    
    # 6. 计算签名 (Signature)
    k_date = hmac_sha256(credential["secret_access_key"].encode("utf-8"), short_x_date)
    k_region = hmac_sha256(k_date, credential["region"])
    k_service = hmac_sha256(k_region, credential["service"])
    k_signing = hmac_sha256(k_service, "request")
    signature = hmac_sha256(k_signing, string_to_sign).hex()

    # 7. 组装最终的 Authorization 请求头
    sign_result["Authorization"] = "HMAC-SHA256 Credential={}, SignedHeaders={}, Signature={}".format(
        credential["access_key_id"] + "/" + credential_scope,
        signed_headers_str,
        signature,
    )
    
    # 8. 发起 HTTP 请求
    # 使用 header.copy() 避免污染外部传入的字典
    final_headers = header.copy() if header else {}
    final_headers.update(sign_result)
    
    response = requests.request(
        method=request_param["method"],
        url="https://{}{}".format(request_param["host"], request_param["path"]),
        headers=final_headers,
        params=request_param["query"],
        data=request_param["body"]
    )
    return response


def main():
    """
    主程序执行入口，用于演示创建 Asset Group 的 API 调用。
    """
    # 准备请求体参数
    data = {
        "Name": "test-group",
        "Description": "test",
        "GroupType": "AIGC",
    }
    data_json = json.dumps(data)
    
    # 获取当前 UTC 时间
    current_date = datetime.datetime.utcnow()
    
    print(f"==================================================")
    print(f"正在发起请求: Action={ACTION}, Version={VERSION}")
    print(f"请求体: {data_json}")
    print(f"==================================================")
    
    # 调用封装好的请求函数
    response = request_api(
        method="POST", 
        date=current_date, 
        query={}, 
        header={}, 
        ak=AK, 
        sk=SK, 
        body=data_json
    )
    
    # 打印响应结果
    print(f"响应状态码: {response.status_code}")
    try:
        # 尝试格式化输出 JSON 响应内容
        parsed_response = response.json()
        print("响应内容:")
        print(json.dumps(parsed_response, indent=4, ensure_ascii=False))
    except json.JSONDecodeError:
        print(f"响应内容:\n{response.text}")

if __name__ == '__main__':
    main()
